Allow passing in hookState from options

fabien opened the pull request:
It can be helpful to pass in `hookState`. Example scenario:

```
Model.observe('before save', function Foo(ctx, next) {
   if (!ctx.hookState.value) return next(); // conditionally run hook
   ctx.data.foo = ctx.hookState.value;
   next();
}); 

Model.updateAll({ ... }, { ... }, { hookState: { value: 'bar' } }, function(err, result) {
    ...
});

```
/cc @raymondfeng @bajtos

fabien commented:
My exact use-case: I'm using `before save` to filter out any unwanted attributes, but there are times where you would want to skip filtering on a case-by-case basis (set: `hookState.skipFilter: true`). Because `options` are not passed in when using the remoting api, it's safe to do this whenever needed.

bajtos commented:
It makes me wonder whether it's better to pass `hookState` from the `options` argument, or whether it's better to pass the `options` argument as `ctx.options`?
/cc @raymondfeng 

@fabien Anyways, the implementation proposed here is ok, please add a unit-test for every method that is being changed/affected.

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -1297,6 +1297,26 @@ module.exports = function(dataSource, should) {
+              where: { id: existingInstance.id },
+              data: { name: 'updated name' },
+              hookState: { foo: 'bar', test: true }
+            }));
```
Don't assert on things that are already verified by other tests, keep the test as focused as possible and assert only on the `hookState`.

```js
observedContexts.hookState.should.eql({ foo: 'bar', test: true });
```

fabien commented:
@bajtos I'll await @raymondfeng opinion - once we decide I will squish and force-push.

Regarding `options` vs. `options.hookState`, I wanted to prevent any clashes with actual `options`.

raymondfeng commented:
I proposed before to make `options` available on the `ctx` object. The `options` can be used to fine control the CRUD, such as transaction. If that happens, there is no need to deal with `options.hookState`.

fabien commented:
@raymondfeng I've changed it to make `options` available to the `ctx` - all tests will check for this now, so it's not needed to add individual unit-tests, similar to when we introduced `hookState`.

raymondfeng commented:
LGTM

bajtos commented:
@crandmck could you please update the doc page "Operations hook" and mention the new context property `ctx.options`?
